### PR TITLE
Feature curved pyramids

### PR DESCRIPTION
**Pull request by jfussbro**
**_Describe your changes here:_**

This PR introduces a cad geometry interpolation algorithm for pyramid elements. With pyramids, t8code now features a cad mapping algorithm for all currently supported element shapes.
The PR also contains a test, which checks all possible linkages of a pyramid against a cad surface and a cad edge.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_cmesh/t8_cmesh_readmshfile.cxx [hunk 1]
```diff
@@ -1017,9 +1017,9 @@ t8_cmesh_msh_file_4_read_eles (t8_cmesh_t cmesh, FILE *fp, sc_hash_t *vertices,
           const t8_geometry_cad_c *cad_geometry = dynamic_cast<const t8_geometry_cad_c *> (cad_geometry_base);
           /* Check for right element class */
           if (eclass != T8_ECLASS_TRIANGLE && eclass != T8_ECLASS_QUAD && eclass != T8_ECLASS_TET
-              && eclass != T8_ECLASS_HEX) {
+              && eclass != T8_ECLASS_HEX && eclass != T8_ECLASS_PYRAMID) {
             t8_errorf (
-              "%s element detected. The cad geometry currently only supports quad, tri, tet and hex elements.\n",
+              "%s element detected. The occ geometry currently only supports quad, tri, tet, hex and pyramid elements.",
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_readmshfile.cxx:1022]:
```suggestion
              "%s element detected. The cad geometry currently only supports quad, tri, tet, hex and pyramid elements.",
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 2]
```diff
@@ -411,6 +411,77 @@ t8_geom_get_scaling_factor_of_edge_on_face_tet (const int edge, const int face,
...
+        { 0, (1 - ref_coords[1]), 0, (1 - ref_coords[0]), 0 },
+        { (1 - ref_coords[1]), 0, 0, (ref_coords[0] - ref_coords[2]), 0 } };
+  const double max_orthogonal_direction[8][5] = { { ref_coords[1], 0, 0, 0, 1 },
+                                                  { 0, ref_coords[1], 0, 0, 1 },
+                                                  { 0, 0, ref_coords[0], 0, 1 },
+                                                  { 0, 0, 0, ref_coords[0], 1 },
+                                                  { (1 - ref_coords[2]), 0, (1 - ref_coords[0]), 0, 0 },
+                                                  { 0, (1 - ref_coords[2]), (1 - ref_coords[2]), 0, 0 },
+                                                  { 0, (1 - ref_coords[2]), 0, (1 - ref_coords[2]), 0 },
+                                                  { (1 - ref_coords[2]), 0, 0, (1 - ref_coords[2]), 0 } };
```
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_helpers.c:446]: Most of the values are calculated and not used

File: src/t8_geometry/t8_geometry_helpers.c [hunk 3]
```diff
@@ -411,6 +411,77 @@ t8_geom_get_scaling_factor_of_edge_on_face_tet (const int edge, const int face,
...
+{
+  /* The function computes the scaling factor of any displacement of a pyramid face, throughout the
+   * volume of the element. The scaling factor is calculated accordingly to 
+   * t8_geom_get_scaling_factor_of_edge_on_face_pyramid with 1 - (orthogonal_direction / max_orthogonal_direction) */
+
+  const double orthogonal_direction[5] = { (ref_coords[0] - ref_coords[2]), (1 - ref_coords[0]),
+                                           (ref_coords[1] - ref_coords[2]), (1 - ref_coords[1]), ref_coords[2] };
+  const double max_orthogonal_direction[5]
+    = { (1 - ref_coords[2]), (1 - ref_coords[2]), (1 - ref_coords[2]), (1 - ref_coords[2]),
+        (ref_coords[0] >= ref_coords[1] ? ref_coords[1] : ref_coords[0]) };
```
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_helpers.c:476]: Same here

File: test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx [hunk 4]
```diff
@@ -889,3 +945,182 @@ INSTANTIATE_TEST_SUITE_P (t8_gtest_check_2d_element_curved_cad_surface, class_2d
...
+        1.0, 1.0, 1.0,    // face vertex 2
+        0.6737526363, 0.5930049972, 0.1275036282 };  // shifted centroid of the pyramid
+
+  double surface_parameters[40]
+    = { 0, 1, 0, 0, 1, 1, 2, 2,   // face 0
+        0, 0, 0, 1, 1, 1, 2, 2,   // face 1
+        0, 0, 0, 1, 1, 1, 2, 2,   // face 2
+        0, 1, 0, 0, 1, 1, 2, 2,   // face 3
+        0, 0, 0, 1, 1, 0, 1, 1 }; // face 4
+    /* The 2's of face 0-3 are placeholders, because these faces only have 3 vertices. */
```
Comment (sandro-elsweijer) [inline test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx:1084]: Use -1 instead, this is how we do it in all other arrays

File: test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx [hunk 5]
```diff
@@ -889,3 +945,182 @@ INSTANTIATE_TEST_SUITE_P (t8_gtest_check_2d_element_curved_cad_surface, class_2d
...
+        1.0, 0.0, 0.0, 1.0, 0.5, 0.0, 1.0, 1.0, 0.0,   // edge 1
+        0.0, 0.0, 0.0, 0.5, 0.0, 0.0, 1.0, 0.0, 0.0,   // edge 2
+        1.0, 1.0, 0.0, 0.5, 1.0, 0.0, 0.0, 1.0, 0.0,   // edge 3
+        1.0, 1.0, 1.0, 0.5, 0.5, 0.5, 0.0, 0.0, 0.0,   // edge 4
+        1.0, 1.0, 1.0, 1.0, 0.5, 0.5, 1.0, 0.0, 0.0,   // edge 5
+        1.0, 1.0, 1.0, 1.0, 1.0, 0.5, 1.0, 1.0, 0.0,   // edge 6
+        1.0, 1.0, 1.0, 0.5, 1.0, 0.5, 0.0, 1.0, 0.0 }; // edge 7
+  double curve_test_return_coords[9]
+    = { 0.0, 0.0, 0.0,    // edge vertex 0
+        0.4999500215, 0.0000523914, 0.4000007585,    // center of edge
```
Comment (sandro-elsweijer) [inline test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx:1107]: Maybe `T8_PRECISION_SQRT_EPS` suffices and we can use 0.5 0 0.4

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_geometry/t8_geometry_helpers.c (hunk 2 above)
Comment (jfussbro) [reply src/t8_geometry/t8_geometry_helpers.c:446]: Every value which is not 0 will get used at some point. Removing the 0's is not possible, since calling the lookup with the edge number wouldn't be possible anymore

File: src/t8_geometry/t8_geometry_helpers.c (hunk 3 above)
Comment (jfussbro) [reply src/t8_geometry/t8_geometry_helpers.c:476]: Why are they not used?

File: test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx (hunk 5 above)
Comment (jfussbro) [reply test/t8_geometry/t8_geometry_implementations/t8_gtest_geometry_cad.cxx:1107]: Nope it doesn't

**Review by Davknapp — CHANGES_REQUESTED** [review]
